<h2>Download and Clean Dataset</h2>



Let's start by importing the pandas and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

<h4>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. 
Ingredients include:</h4>

<h4>1. Cement</h4>

<h4>2. Blast Furnace Slag</h4>

<h4>3. Fly Ash</h4>

<h4>4. Water</h4>

<h4>5. Superplasticizer</h4>

<h4>6. Coarse Aggregate</h4>

<h4>7. Fine Aggregate</h4>

Let's read the dataset into a pandas dataframe.

In [2]:
     concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 
162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine 
aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.


<h4>Let's check how many data points we have.</h4>

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit 
the training data.

Let's check the dataset for any missing values.

In [4]:

concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

<h4>Split data into predictors and target</h4>



The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

<h2>Import Keras</h2>


<h4>Let's go ahead and import the Keras library</h4>

In [11]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.



In [12]:
from keras.models import Sequential
from keras.layers import Dense

In [13]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has three hidden layers each with 10 neurons and ReLU activation function. 
It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [14]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [15]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

<h2>Train and Test the Network</h2>


Let's call the function now to create our model.

In [16]:

# build the model
model = regression_model()

Next, we will train the model for 50 epochs.



In [17]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)

Epoch 1/50
23/23 - 0s - loss: 1576.5809
Epoch 2/50
23/23 - 0s - loss: 1540.0042
Epoch 3/50
23/23 - 0s - loss: 1485.6138
Epoch 4/50
23/23 - 0s - loss: 1400.1674
Epoch 5/50
23/23 - 0s - loss: 1268.0570
Epoch 6/50
23/23 - 0s - loss: 1079.1414
Epoch 7/50
23/23 - 0s - loss: 832.6203
Epoch 8/50
23/23 - 0s - loss: 579.6337
Epoch 9/50
23/23 - 0s - loss: 383.5942
Epoch 10/50
23/23 - 0s - loss: 291.7711
Epoch 11/50
23/23 - 0s - loss: 256.8469
Epoch 12/50
23/23 - 0s - loss: 239.2166
Epoch 13/50
23/23 - 0s - loss: 225.9625
Epoch 14/50
23/23 - 0s - loss: 217.6028
Epoch 15/50
23/23 - 0s - loss: 210.4634
Epoch 16/50
23/23 - 0s - loss: 203.5963
Epoch 17/50
23/23 - 0s - loss: 198.7607
Epoch 18/50
23/23 - 0s - loss: 192.8550
Epoch 19/50
23/23 - 0s - loss: 189.1631
Epoch 20/50
23/23 - 0s - loss: 184.0285
Epoch 21/50
23/23 - 0s - loss: 180.4102
Epoch 22/50
23/23 - 0s - loss: 176.6122
Epoch 23/50
23/23 - 0s - loss: 172.9960
Epoch 24/50
23/23 - 0s - loss: 169.7109
Epoch 25/50
23/23 - 0s - loss: 166.6423
Epo

Next we need to evaluate the model on the test data.

In [18]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 698us/step - loss: 129.4819


129.48193359375

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [19]:
from sklearn.metrics import mean_squared_error

In [20]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

129.48193612938482 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [21]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 101.44290161132812
MSE 2: 115.58789825439453
MSE 3: 87.2984390258789
MSE 4: 60.804161071777344
MSE 5: 44.8884162902832
MSE 6: 45.33283233642578
MSE 7: 43.32044982910156
MSE 8: 32.40530014038086
MSE 9: 35.277584075927734
MSE 10: 37.89807891845703
MSE 11: 32.99872970581055
MSE 12: 27.64239501953125
MSE 13: 37.7236328125
MSE 14: 36.98420333862305
MSE 15: 28.435888290405273
MSE 16: 25.67913055419922
MSE 17: 28.053436279296875
MSE 18: 28.187885284423828
MSE 19: 25.423824310302734
MSE 20: 29.43933868408203
MSE 21: 27.753103256225586
MSE 22: 24.329315185546875
MSE 23: 19.80491828918457
MSE 24: 25.47101593017578
MSE 25: 28.540761947631836
MSE 26: 30.46794319152832
MSE 27: 22.535364151000977
MSE 28: 24.51779556274414
MSE 29: 25.961896896362305
MSE 30: 24.325733184814453
MSE 31: 19.67184066772461
MSE 32: 22.362346649169922
MSE 33: 22.38640022277832
MSE 34: 22.90465545654297
MSE 35: 28.004594802856445
MSE 36: 28.314165115356445
MSE 37: 18.819095611572266
MSE 38: 24.732208251953125
MSE 39: 